lib/approval_request_listener.rb

```diff
@@ -0,0 +1,64 @@
+require "manageiq-messaging"
+
+class ApprovalRequestListener
+  SERVICE_NAME = "platform.approval".freeze
+  CLIENT_AND_GROUP_REF = "approval-catalog-api-worker".freeze
+  EVENT_REQUEST_FINISHED = 'request_finished'.freeze
+
+  attr_accessor :messaging_client_options, :client
+
+  def initialize(messaging_client_options = {})
+    self.messaging_client_options = default_messaging_options.merge(messaging_client_options)
+  end
+
+  def run
+    Thread.new { subscribe_to_approval_updates }
+  end
+
+  def subscribe_to_approval_updates
+    self.client = ManageIQ::Messaging::Client.open(messaging_client_options)
+
+    client.subscribe_topic(
+      :service   => SERVICE_NAME,
+      :max_bytes => 500_000
+    ) do |messages|
+      messages.each do |msg|
+        process_message(msg)
+      end
+    end
+  ensure
+    client&.close
+    self.client = nil
+  end
+
+  private
+
+  def process_message(msg)
+    approval = ApprovalRequest.find_by!(:approval_request_ref => msg.payload["request_id"])
+    approval.order_item.update_message("info", "Task update message received with payload: #{msg.payload}")
+    if msg.message == EVENT_REQUEST_FINISHED && (msg.payload["decision"] == "approved" || msg.payload["decision"] == "denied")
+      update_and_log_state(approval, msg.payload)
+    end
+  rescue ActiveRecord::RecordNotFound
+    ProgressMessage.create!(
```

Comment:
@syncrou This either needs to be a error log and not a progress message, if we didn't find the OrderItem or the ApprovalRequest.